src/domain/core/src/utils/time_source.rs

```diff
@@ -24,46 +28,132 @@ pub trait SystemTimeSource: Send + Sync {
 #[dill::interface(dyn SystemTimeSource)]
 pub struct SystemTimeSourceDefault;
 
+#[async_trait::async_trait]
 impl SystemTimeSource for SystemTimeSourceDefault {
     fn now(&self) -> DateTime<Utc> {
         Utc::now()
     }
+
+    async fn wake_after(&self, duration: Duration) {
+        let std_duration = duration.to_std().unwrap();
+
+        tokio::time::sleep(std_duration).await;
+    }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+type AwaitingCallerId = usize;
+type WokeUpCallersIds = Vec<AwaitingCallerId>;
+
+#[derive(Debug)]
+struct AwaitingCaller {
+    id: AwaitingCallerId,
+    time_to_wake_up: Duration,
```

Comment:
It's not a "time", it's an offset from `t`. If the idea was to store time, the type should not be `Duration`.

Reply:
fixed